drivers/vdpa/virtio/virtio_vdpa.c

```diff
@@ -0,0 +1,955 @@
+{
+	int ret = -EAGAIN;
+	struct rte_intr_handle *intr_handle;
+	int retries = VIRTIO_VDPA_INTR_RETRIES;
+
+	intr_handle = priv->vrings[index]->intr_handle;
+	if (rte_intr_fd_get(intr_handle) != -1) {
+		while (retries-- && ret == -EAGAIN) {
+			ret = rte_intr_callback_unregister(intr_handle,
+							virtio_vdpa_virtq_handler,
+							priv->vrings[index]);
+			if (ret == -EAGAIN) {
+				DRV_LOG(DEBUG, "Try again to unregister fd %d "
+				"of virtq %d interrupt, retries = %d.",
+				rte_intr_fd_get(intr_handle),
+				(int)priv->vrings[index]->index, retries);
+				usleep(VIRTIO_VDPA_INTR_RETRIES_USEC);
+			}
+		}
+		rte_intr_fd_set(intr_handle, -1);
+	}
+	rte_intr_instance_free(intr_handle);
+	return 0;
+}
+
+static int
+virtio_vdpa_virtq_doorbell_relay_enable(struct virtio_vdpa_priv *priv, int index)
+{
+	int ret;
+	struct rte_vhost_vring vq;
+	struct rte_intr_handle *intr_handle;
+
+	ret = rte_vhost_get_vhost_vring(priv->vid, index, &vq);
+	if (ret)
+		return ret;
+
+	intr_handle = rte_intr_instance_alloc(RTE_INTR_INSTANCE_F_SHARED);
+	if (intr_handle == NULL) {
+		DRV_LOG(ERR, "Fail to allocate intr_handle");
+		return -EINVAL;
+	}
+
+	priv->vrings[index]->intr_handle = intr_handle;
+	if (rte_intr_fd_set(intr_handle, vq.kickfd)) {
+		DRV_LOG(ERR, "Fail to set kick fd");
+		goto error;
+	}
+
+	if (rte_intr_fd_get(intr_handle) == -1) {
+		DRV_LOG(ERR, "Virtq %d kickfd is invalid.", index);
+		goto error;
+	} else {
+		if (rte_intr_type_set(intr_handle, RTE_INTR_HANDLE_EXT))
+			goto error;
+
+		if (rte_intr_callback_register(intr_handle,
+						   virtio_vdpa_virtq_handler,
+						   priv->vrings[index])) {
+			rte_intr_fd_set(intr_handle, -1);
+			DRV_LOG(ERR, "Failed to register virtq %d interrupt.",
+				index);
+			goto error;
+		} else {
+			DRV_LOG(DEBUG, "Register fd %d interrupt for virtq %d.",
+				rte_intr_fd_get(intr_handle),
+				index);
+		}
+	}
+
+	return 0;
+error:
+	virtio_vdpa_virtq_doorbell_relay_disable(priv, index);
+	return -EINVAL;
+}
+
+static int
+virtio_vdpa_virtq_disable(struct virtio_vdpa_priv *priv, int index)
+{
+	int ret;
+
+	ret = virtio_vdpa_virtq_doorbell_relay_disable(priv, index);
+	if (ret) {
+		 DRV_LOG(ERR, "Virtq doorbell relay failed");
+		 return ret;
+	}
+
+	virtio_pci_dev_queue_del(priv->vpdev, index);
+
+	priv->vrings[index]->notifier_state = VIRTIO_VDPA_NOTIFIER_STATE_DISABLED;
+
+	ret = virtio_pci_dev_interrupts_disable(priv->vpdev, index+1);
+	if (ret) {
+		 DRV_LOG(ERR, "Virtq %d interrupt disabel failed", index);
+		 return ret;
+	}
+
+	priv->vrings[index]->enable = 0;
+	return 0;
+}
+
+static int
+virtio_vdpa_virtq_enable(struct virtio_vdpa_priv *priv, int index)
+{
+	int ret;
+	int vid;
+	struct rte_vhost_vring vq;
+	struct virtio_pci_dev_vring_info vring_info;
+	uint64_t gpa;
+
+	vid = priv->vid;
+
+	ret = rte_vhost_get_vhost_vring(vid, index, &vq);
+	if (ret)
+		return ret;
+
+	ret =virtio_pci_dev_interrupts_enable(priv->vpdev, vq.callfd, index+1);
```

Comment:
this should be done only one, or its name should be pci_dev_interrupt_enable

Reply:
interrupt is enabled per queue, so, i just need to change to pci_dev_interrupt_enable?